### ISSUE TITLE
Block expression should warning about missing return value

### ISSUE THREAD

**Issue by hfiguiere** [rust-lang/rust#26577]
This is part of the Rust tutorial session. The whole purpose of the exercise was to find a mistake in the block expression.

See Exercise 2.1
http://pnkfelix.github.io/cyot/tutorial/exercises/ex_part_2

With rust compile this:

``` rust
fn foo() -> i32 {
   42
}

pub fn main() {

    let the_sum :i32 = {
        foo();
    };

    let the_other_sum = {
        foo();
    };

}
```

It outputs this error:

```
bug_block.rs:7:24: 9:6 error: mismatched types:
 expected `i32`,
    found `()`
(expected i32,
    found ()) [E0308]
bug_block.rs:7     let the_sum :i32 = {
bug_block.rs:8         foo();
bug_block.rs:9     };
error: aborting due to previous error
```

It should also put a warning for line 12: "possibly missing return value". This would catch the bug.

**Comment by steveklabnik** [comment]
Possibly related to https://github.com/rust-lang/rust/issues/26576

**Comment by hfiguiere** [comment]
It was found during the same thing. My understanding of what's happening is that the return type is being inferred so it assume that returning '()' is valid. Since I get the error when I actually specify a type for the block expression.

**Comment by hfiguiere** [comment]
Here is my suggestion:

If the type of the block expression is inferred and there no return value, emit a warning.

Suggested warning removal: specify an explicit return value as
````
let the_other_sum = {
    foo();
    ()
};
````

**Comment by arielb1** [comment]
@hfiguiere 

lints run after type-checking, so this won't help.

**Comment by pnkfelix** [comment]
Another, more heavy handed approach: During parsing, if a block expression occurs in an syntactic context that is expecting a value, then do the warning (at parse time) suggesting that one put in an explicit return value.

So:

```rust
fn quux() {
  {
    hi();
  } // this is fine

  let x = {
    foo();
  }; // this causes the warning

  bar({
    foo();
  }); // this causes the warning too
}
```

**Comment by memoryruins** [comment]
Triage: the error message has improved
```
error[E0308]: mismatched types
 --> src/main.rs:7:24
  |
7 |       let the_sum :i32 = {
  |  ________________________^
8 | |         foo();
  | |              - help: consider removing this semicolon
9 | |     };
  | |_____^ expected i32, found ()
  |
  = note: expected type `i32`
             found type `()`

error: aborting due to previous error
```
`rustc: 1.32.0` / `rust version 1.34.0-nightly (c1c3c4e95 2019-01-29)`

**Comment by hfiguiere** [comment]
Yes this is so much better.

**Comment by arielb1** [comment]
Fixed in #57381, so soft-duplicate of #57348. I think it's better to add a test for the non-if-then-else case.
